File: frontend/src/shared/api/catalog.ts
import { apiRequest } from './client'

export interface Make {
  id: number
  name: string
  country: string
  brand_tier: string
}

export interface Model {
  id: number
  make_id: number
  name: string
  segment: string
  body_type: string
}

export interface Generation {
  id: number
  model_id: number
  name: string
  year_from: number
  year_to: number | null
  restyling: number
}

export interface Modification {
  id: number
  make: string
  model: string
  generation: string
  trim_name: string | null
  year_from: number
  year_to: number | null
  body_type: string
  segment: string
  power_hp: number
  fuel_type: string
  transmission: string
  drive: string
  fuel_consumption_combined_l_100km: number
  msrp_new_rub: number
}

export interface Region {
  id: number
  name: string
  iso_code: string
  federal_district: string
  climate_zone: string
}

export const catalogApi = {
  listMakes: (query: string) =>
    apiRequest<Make[]>(`/api/catalog/makes?q=${encodeURIComponent(query)}`),
  listRegions: () => apiRequest<Region[]>('/api/catalog/regions'),
  listModelsByMake: (makeId: number) => apiRequest<Model[]>(`/api/catalog/makes/${makeId}/models`),
  listGenerationsByModel: (modelId: number) =>
    apiRequest<Generation[]>(`/api/catalog/models/${modelId}/generations`),
  listModificationsByGeneration: (generationId: number) =>
    apiRequest<Modification[]>(`/api/catalog/generations/${generationId}/modifications`),
  getModification: (modificationId: number) =>
    apiRequest<Modification>(`/api/catalog/modifications/${modificationId}`),
}

File: backend/app/api/schemas/catalog.py
"""Pydantic v2 схемы caталог-эндпоинтов."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class RegionRead(BaseModel):
    """Регион РФ для выпадающего списка в UI."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    iso_code: str
    federal_district: str
    climate_zone: str


class MakeRead(BaseModel):
    """Марка автомобиля."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    country: str
    brand_tier: str = Field(
        description="russian | mass | japanese_korean_mass | premium | chinese"
    )


class ModelRead(BaseModel):
    """Модель в рамках марки."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    make_id: int
    name: str
    segment: str = Field(description="A | B | C | D | E | F | J_SUV | J_CROSS | M_MPV | LCV | S_SPORT")
    body_type: str


class GenerationRead(BaseModel):
    """Поколение модели."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    model_id: int
    name: str
    year_from: int
    year_to: int | None = None
    restyling: int = 0


class ModificationRead(BaseModel):
    """Slim-карточка модификации."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    make: str
    model: str
    generation: str
    trim_name: str | None = None
    year_from: int
    year_to: int | None = None
    body_type: str
    segment: str
    power_hp: int
    engine_volume_l: float | None = None
    fuel_type: str
    transmission: str
    drive: str
    fuel_consumption_combined_l_100km: float
    msrp_new_rub: int


__all__ = [
    "RegionRead",
    "MakeRead",
    "ModelRead",
    "GenerationRead",
    "ModificationRead",
]
